packages/wdio-sauce-service/src/launcher.js

```diff
@@ -1,59 +1,75 @@
+    }
+
     /**
      * modify config and launch sauce connect
      */
-    onPrepare (config, capabilities) {
-        if (!config.sauceConnect) {
+    async onPrepare (config, capabilities) {
+        if (!this.options.sauceConnect) {
             return
         }
 
+        const sauceConnectOpts = Object.assign({ noAutodetect: true }, this.options.sauceConnectOpts)
+        const sauceConnectTunnelIdentifier = (
+            sauceConnectOpts.tunnelIdentifier ||
+            /**
+             * generate random identifier if not provided
+             */
+            `SC-tunnel-${Math.random().toString().slice(2)}`)
+
         this.sauceConnectOpts = Object.assign({
             username: config.user,
             accessKey: config.key,
-            logger: log.debug
-        }, config.sauceConnectOpts)
+            logger: log.debug,
+            tunnelIdentifier: sauceConnectTunnelIdentifier
+        }, sauceConnectOpts)
 
-        if (config.scRelay) {
+        if (this.options.scRelay) {
             config.protocol = 'http'
             config.hostname = 'localhost'
             config.port = this.sauceConnectOpts.port || 4445
         }
 
-        const sauceConnectTunnelIdentifier = this.sauceConnectOpts.tunnelIdentifier
-
-        if (sauceConnectTunnelIdentifier) {
-            if (Array.isArray(capabilities)) {
-                capabilities.forEach(capability => {
-                    if (capability['sauce:options'] === undefined) {
-                        capability.tunnelIdentifier = capability.tunnelIdentifier || sauceConnectTunnelIdentifier
-                    } else {
-                        capability['sauce:options'].tunnelIdentifier = capability['sauce:options'].tunnelIdentifier || sauceConnectTunnelIdentifier
-                    }
-                })
-            } else {
-                Object.keys(capabilities).forEach(browser => {
-                    if (capabilities[browser].capabilities['sauce:options'] === undefined) {
-                        capabilities[browser].capabilities.tunnelIdentifier = capabilities[browser].capabilities.tunnelIdentifier || sauceConnectTunnelIdentifier
-                    } else {
-                        capabilities[browser].capabilities['sauce:options'].tunnelIdentifier = capabilities[browser].capabilities['sauce:options'].tunnelIdentifier || sauceConnectTunnelIdentifier
-                    }
-                })
-            }
+        if (Array.isArray(capabilities)) {
+            capabilities.forEach(capability => {
+                if (capability['sauce:options'] === undefined) {
+                    capability.tunnelIdentifier = capability.tunnelIdentifier || sauceConnectTunnelIdentifier
+                } else {
+                    capability['sauce:options'].tunnelIdentifier = capability['sauce:options'].tunnelIdentifier || sauceConnectTunnelIdentifier
+                }
+            })
+        } else {
+            Object.keys(capabilities).forEach(browser => {
```

Comment:
This is just a code style suggestion, but in some other services we're using `for..in` loops and here we're using a `forEach`, maybe we should add a linter to force one or the other?

Reply:
replaced it with for..of loop. I don't have a strong opinion on enforcing it. This should be a separated issue/PR then as I think that there are a bunch of forEach loops used in the code base.